From 5ddcce538067f16309368f8c55caba1768193405 Mon Sep 17 00:00:00 2001
From: Kevin Wolf <kwolf@redhat.com>
Date: Wed, 26 May 2010 13:38:13 -0300
Subject: [PATCH 13/20] qdev-properties: Fix (u)intXX parsers

RH-Author: Kevin Wolf <kwolf@redhat.com>
Message-id: <1274881093-26849-1-git-send-email-kwolf@redhat.com>
Patchwork-id: 9579
O-Subject: [RHEL-6 qemu-kvm PATCH] qdev-properties: Fix (u)intXX parsers
Bugzilla: 596093
RH-Acked-by: Gleb Natapov <gleb@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Gerd Hoffmann <kraxel@redhat.com>

Bugzilla: 596093
Upstream status: Submitted

scanf calls must not use PRI constants, they have probably the wrong size and
corrupt memory. We could replace them by SCN ones, but strtol is simpler than
scanf here anyway. While at it, also fix the parsers to reject garbage after
the number ("4096xyz" was accepted before).

Signed-off-by: Kevin Wolf <kwolf@redhat.com>
---
 hw/qdev-properties.c |   50 +++++++++++++++++++++++++++++++++++---------------
 1 files changed, 35 insertions(+), 15 deletions(-)

Signed-off-by: Eduardo Habkost <ehabkost@redhat.com>
---
 hw/qdev-properties.c |   50 +++++++++++++++++++++++++++++++++++---------------
 1 files changed, 35 insertions(+), 15 deletions(-)

diff --git a/hw/qdev-properties.c b/hw/qdev-properties.c
index 9ffdba7..64d83ce 100644
--- a/hw/qdev-properties.c
+++ b/hw/qdev-properties.c
@@ -68,12 +68,14 @@ PropertyInfo qdev_prop_bit = {
 static int parse_uint8(DeviceState *dev, Property *prop, const char *str)
 {
     uint8_t *ptr = qdev_get_prop_ptr(dev, prop);
-    const char *fmt;
+    char *end;
 
     /* accept both hex and decimal */
-    fmt = strncasecmp(str, "0x",2) == 0 ? "%" PRIx8 : "%" PRIu8;
-    if (sscanf(str, fmt, ptr) != 1)
+    *ptr = strtoul(str, &end, 0);
+    if ((*end != '\0') || (end == str)) {
         return -EINVAL;
+    }
+
     return 0;
 }
 
@@ -96,12 +98,14 @@ PropertyInfo qdev_prop_uint8 = {
 static int parse_uint16(DeviceState *dev, Property *prop, const char *str)
 {
     uint16_t *ptr = qdev_get_prop_ptr(dev, prop);
-    const char *fmt;
+    char *end;
 
     /* accept both hex and decimal */
-    fmt = strncasecmp(str, "0x",2) == 0 ? "%" PRIx16 : "%" PRIu16;
-    if (sscanf(str, fmt, ptr) != 1)
+    *ptr = strtoul(str, &end, 0);
+    if ((*end != '\0') || (end == str)) {
         return -EINVAL;
+    }
+
     return 0;
 }
 
@@ -124,12 +128,14 @@ PropertyInfo qdev_prop_uint16 = {
 static int parse_uint32(DeviceState *dev, Property *prop, const char *str)
 {
     uint32_t *ptr = qdev_get_prop_ptr(dev, prop);
-    const char *fmt;
+    char *end;
 
     /* accept both hex and decimal */
-    fmt = strncasecmp(str, "0x",2) == 0 ? "%" PRIx32 : "%" PRIu32;
-    if (sscanf(str, fmt, ptr) != 1)
+    *ptr = strtoul(str, &end, 0);
+    if ((*end != '\0') || (end == str)) {
         return -EINVAL;
+    }
+
     return 0;
 }
 
@@ -150,9 +156,13 @@ PropertyInfo qdev_prop_uint32 = {
 static int parse_int32(DeviceState *dev, Property *prop, const char *str)
 {
     int32_t *ptr = qdev_get_prop_ptr(dev, prop);
+    char *end;
 
-    if (sscanf(str, "%" PRId32, ptr) != 1)
+    *ptr = strtol(str, &end, 10);
+    if ((*end != '\0') || (end == str)) {
         return -EINVAL;
+    }
+
     return 0;
 }
 
@@ -175,9 +185,13 @@ PropertyInfo qdev_prop_int32 = {
 static int parse_hex32(DeviceState *dev, Property *prop, const char *str)
 {
     uint32_t *ptr = qdev_get_prop_ptr(dev, prop);
+    char *end;
 
-    if (sscanf(str, "%" PRIx32, ptr) != 1)
+    *ptr = strtoul(str, &end, 16);
+    if ((*end != '\0') || (end == str)) {
         return -EINVAL;
+    }
+
     return 0;
 }
 
@@ -200,12 +214,14 @@ PropertyInfo qdev_prop_hex32 = {
 static int parse_uint64(DeviceState *dev, Property *prop, const char *str)
 {
     uint64_t *ptr = qdev_get_prop_ptr(dev, prop);
-    const char *fmt;
+    char *end;
 
     /* accept both hex and decimal */
-    fmt = strncasecmp(str, "0x",2) == 0 ? "%" PRIx64 : "%" PRIu64;
-    if (sscanf(str, fmt, ptr) != 1)
+    *ptr = strtoull(str, &end, 0);
+    if ((*end != '\0') || (end == str)) {
         return -EINVAL;
+    }
+
     return 0;
 }
 
@@ -228,9 +244,13 @@ PropertyInfo qdev_prop_uint64 = {
 static int parse_hex64(DeviceState *dev, Property *prop, const char *str)
 {
     uint64_t *ptr = qdev_get_prop_ptr(dev, prop);
+    char *end;
 
-    if (sscanf(str, "%" PRIx64, ptr) != 1)
+    *ptr = strtoull(str, &end, 16);
+    if ((*end != '\0') || (end == str)) {
         return -EINVAL;
+    }
+
     return 0;
 }
 
-- 
1.7.0.3

